crates/ruff_python_parser/src/string.rs

```diff
@@ -173,319 +155,29 @@ impl<'a> StringParser<'a> {
-        let mut values = vec![];
-
-        while let Some(ch) = self.peek() {
-            match ch {
-                '{' => {
-                    if nested == 0 {
-                        match self.peek2() {
-                            Some('{') => {
-                                self.next_char();
-                                self.next_char();
-                                content.push('{');
-                                continue;
-                            }
-                            None => {
-                                return Err(FStringError::new(UnclosedLbrace, self.get_pos()).into())
-                            }
-                            _ => {}
-                        }
-                    }
-                    if !content.is_empty() {
-                        values.push(Expr::from(ast::ExprConstant {
-                            value: std::mem::take(&mut content).into(),
-                            kind: None,
-                            range: self.range(start_location),
-                        }));
-                    }
-
-                    let parsed_values = self.parse_formatted_value(nested)?;
-                    values.extend(parsed_values);
-                    start_location = self.get_pos();
-                }
-                '}' => {
-                    if nested > 0 {
-                        break;
-                    }
-                    self.next_char();
-                    if let Some('}') = self.peek() {
-                        self.next_char();
-                        content.push('}');
-                    } else {
-                        return Err(FStringError::new(SingleRbrace, self.get_pos()).into());
-                    }
-                }
                 '\\' if !self.kind.is_raw() => {
-                    self.next_char();
-                    content.push_str(&self.parse_escaped_char()?);
+                    value.push_str(&self.parse_escaped_char()?);
                 }
-                _ => {
-                    content.push(ch);
-                    self.next_char();
+                // If there are any curly braces inside a `FStringMiddle` token,
+                // then they were escaped (i.e. `{{` or `}}`). This means that
+                // we need increase the location by 2 instead of 1.
+                ch @ ('{' | '}') => {
+                    self.location += ch.text_len();
+                    value.push(ch);
                 }
+                ch => value.push(ch),
             }
         }
-
-        if !content.is_empty() {
-            values.push(Expr::from(ast::ExprConstant {
-                value: content.into(),
-                kind: None,
-                range: self.range(start_location),
-            }));
-        }
-
-        Ok(values)
+        Ok(Expr::from(ast::ExprConstant {
+            value: value.into(),
+            kind: self.kind.is_unicode().then(|| "u".to_string()),
```

Comment:
Lol, we should change kind to an enum. Allocating a string for this is silly (and I'm surprised the kind is a flag on constant when only strings can be unicode). But we can do this as a separate PR.

Reply:
I think it should be a boolean field `unicode` similar to `implicit_concatenated` on `StringConstant`. What do you think?